Experiment - 12 -> Neymeni test on Iris Dataset

In [ ]:
pip install scikit-posthocs

In [ ]:
# Importing libraries
from scipy import stats
import scikit_posthocs as sp
import numpy as np
import pandas as pd

In [ ]:
df = pd.read_csv('/content/Iris_Data.csv')

In [ ]:
# Data groups
sepal_len_setosa = df.loc[df['species']=='Iris-setosa', 'sepal_length'].to_numpy()
sepal_len_versicolor = df.loc[df['species']=='Iris-versicolor', 'sepal_length'].to_numpy()
sepal_len_virginica = df.loc[df['species']=='Iris-virginica', 'sepal_length'].to_numpy()

In [ ]:
sepal_len_setosa

array([5.1, 4.9, 4.7, 4.6, 5. , 5.4, 4.6, 5. , 4.4, 4.9, 5.4, 4.8, 4.8,
       4.3, 5.8, 5.7, 5.4, 5.1, 5.7, 5.1, 5.4, 5.1, 4.6, 5.1, 4.8, 5. ,
       5. , 5.2, 5.2, 4.7, 4.8, 5.4, 5.2, 5.5, 4.9, 5. , 5.5, 4.9, 4.4,
       5.1, 5. , 4.5, 4.4, 5. , 5.1, 4.8, 5.1, 4.6, 5.3, 5. ])

In [ ]:
sepal_len_versicolor

array([7. , 6.4, 6.9, 5.5, 6.5, 5.7, 6.3, 4.9, 6.6, 5.2, 5. , 5.9, 6. ,
       6.1, 5.6, 6.7, 5.6, 5.8, 6.2, 5.6, 5.9, 6.1, 6.3, 6.1, 6.4, 6.6,
       6.8, 6.7, 6. , 5.7, 5.5, 5.5, 5.8, 6. , 5.4, 6. , 6.7, 6.3, 5.6,
       5.5, 5.5, 6.1, 5.8, 5. , 5.6, 5.7, 5.7, 6.2, 5.1, 5.7])

In [ ]:
sepal_len_virginica

array([6.3, 5.8, 7.1, 6.3, 6.5, 7.6, 4.9, 7.3, 6.7, 7.2, 6.5, 6.4, 6.8,
       5.7, 5.8, 6.4, 6.5, 7.7, 7.7, 6. , 6.9, 5.6, 7.7, 6.3, 6.7, 7.2,
       6.2, 6.1, 6.4, 7.2, 7.4, 7.9, 6.4, 6.3, 6.1, 7.7, 6.3, 6.4, 6. ,
       6.9, 6.7, 6.9, 5.8, 6.8, 6.7, 6.7, 6.3, 6.5, 6.2, 5.9])

In [ ]:
 
# Conduct the Friedman Test
stats.friedmanchisquare(sepal_len_setosa, sepal_len_versicolor, sepal_len_virginica)
 
# Combine three groups into one array
data = np.array([sepal_len_setosa, sepal_len_versicolor, sepal_len_virginica])
 
# Conduct the Nemenyi post-hoc test
sp.posthoc_nemenyi_friedman(data.T)

0         1         2
0  1.000  0.001000  0.001000
1  0.001  1.000000  0.001951
2  0.001  0.001951  1.000000

We need to transpose the NumPy array (data.T) to conduct the post-hoc test efficiently. As you can see in the output, The Nemenyi post-hoc test produces the p-values for each pairwise comparison of means. These values are:

P-value of group 0 vs. group 1: 0.001
P-value of group 0 vs. group 2: 0.001
P-value of group 1 vs. group 2: 0.001951
For α = 0.05 all three groups seem to have statistically significantly different means, which is true since sepal lengths of three different types of flowers were taken into consideration